Doc note for complex

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41252 Doc note for complex**

Differential Revision: [D22553266](https://our.internmc.facebook.com/intern/diff/D22553266)

anjali411 committed:
doc note for complex

[ghstack-poisoned]

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,155 @@
+
+Users who currently worked around the lack of complex tensors with real tensors of shape `(..., 2)`
+can easily to switch using the complex tensors in their code using :func:`torch.view_as_complex` and
+- :func:`torch.view_as_real`:
```
I'd suggest mentioning that this involves ZERO copying

anjali411 replied:
done!

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,155 @@
+Autograd Support (with and without view_as_real, view_as_complex)
+-----------------------------------------------------------------
+
+
```
here might be a good places to talk about holomorphic functions. Maybe the "upshot" section of https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#Complex-numbers-and-differentiation

anjali411 replied:
should we just add it to the autograd doc and link that doc here?

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,155 @@
+JIT
+distributions
+
+multiprocessing
```
I don't see any reason why this wouldn't work

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,155 @@
+
+multiprocessing
+
+distributed
```
If distributed doesn't work, it doesn't for any good reason, and we should fix it (shouldn't be hard)

anjali411 replied:
Synced with @mrshenli and we will have to add mappings between complex and float dtypes for data transfer (because for eg. [nccl](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/types.html?highlight=ncclfloat) doesn't support complex dtypes) and add some tests to ensure that it works correctly. These changes will be included as part of `1.7` because it's too last minute to add it in `1.6`.

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,155 @@
+by using Float Tensors with shape :math:`(..., 2)` where the last dimension contains the real and imaginary values.
+
+We support many functions for Complex Tensors eg. SVD, QR etc. Operations on complex tensors are likely to be
+faster than operations on float tensors mimicking them. We also have a `Vec256 class` for `ComplexFloat` and
```
Vec256 is a bit internal and I probably wouldn't mention it in a user doc like this

gchanan replied:
Also `ComplexFloat` and `ComplexDouble` haven't been defined yet, but they also sound internal.

You can just say something like "Operations involving complex numbers in PyTorch are optimized to use vectorized assembly instructions and specialized kernels (e.g. LAPACK, CuBlas)."

I'm not really sure why you want to compare users writing their own functions -- why would they do that?

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,155 @@
+Complex Numbers
+===============
+
+Complex Numbers are numbers that can be expressed in the form :math:`a + bi`, where a and b are real numbers,
```
Don't capitalize complex numbers

anjali411 committed:
Update on "doc note for complex"




[ghstack-poisoned]

gchanan commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,174 @@
+
+Complex Numbers are numbers that can be expressed in the form :math:`a + bi`, where a and b are real numbers,
+and *i* is a solution of the equation :math:`x^2 = −1`. Complex numbers frequently occur in mathematics and
+engineering, especially in signal processing. The aim of introducing Complex Tensors is to provide a more natural
```
nit: this sentence describes the current state, but it's going to be out of date soon after you write it (i.e., we'll have complex support and torchaudio won't have hacks.  You can just say something like "Complex Tensors are provided..."

gchanan replied:
nit: it's a little wordier, but "Tensors of complex dtype" is more descriptive/accurate.

gchanan replied:
nit: I believe "workaround" is two words when used as a verb.

gchanan commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,174 @@
+
+<add code>
+
+The following factory functions can be used to create complex tensors:
```
are there any factory functions that *dont* support complex tensors?

anjali411 replied:
`torch.arange`, `torch.range`, `torch.linspace`, `torch.logspace`, `torch.quantize_per_tensor`, `torch.quantize_per_channel`

do you think it's better to just have a list of the factory functions not supported instead?

gchanan commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,174 @@
+Autograd
+--------
+
+PyTorch supports Autograd for Complex Numbers.
```
I think the core of this should go in the autograd section and you should link it from here.

anjali411 replied:
so should we move everything mentioned here (including the example) to autograd mechanics as well? 
cc. @albanD

albanD replied:
I think it is already in the autograd doc as well.
The list explaining what you get for different type of functions is good to have here I think.
Not sure what the code sample brings though.

gchanan commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,174 @@
+distributions, multiprocessing, distributed, ONNX??
+
+If any of these would help your use case, please search if an issue has already been filed (https://github.com/pytorch/pytorch/issues?q=is%3Aopen+is%3Aissue+label%3A%22module%3A+named+tensor%22)
+and if not, file one(https://github.com/pytorch/pytorch/issues/new/choose).
```
nit: space

gchanan commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,174 @@
+          [[2., 2.],
+          [2., 2.]]], dtype=torch.float64)
+
+We also do not support the following subsystems:
```
"also" is a little strange here, as this doesn't follow the previous sections (which are things that are supported).

anjali411 replied:
yeah you are right updated!

anjali411 committed:
Update on "[WIP] Doc note for complex"




[ghstack-poisoned]

albanD commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,188 @@
+Complex Numbers
+===============
+
+Complex Numbers are numbers that can be expressed in the form :math:`a + bi`, where a and b are real numbers,
```
Note that you use `j` in the autograd part of the doc. Maybe you want to make that consistent.

albanD commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,188 @@
+
+Users who currently worked around the lack of complex tensors with real tensors of shape `(..., 2)`
+can easily to switch using the complex tensors in their code using :func:`torch.view_as_complex` and
+- :func:`torch.view_as_real` view functions:
```
I'm not sure that just saying "view functions" makes it clear to the user that these are views. Maybe you want to insist on it a bit more as it is fairly important.

anjali411 replied:
`...:func:`torch.view_as_complex` and - :func:`torch.view_as_real` functions which involve zero copy.` does that look better?

albanD replied:
Or add a new setence: Note that these functions don't perform any copy and return a view of the input Tensor.

albanD commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,188 @@
+(the latter three are only supported on CPU). However we are working to add support for more
+functions soon: :func:`torch.matmul`, :func:`torch.solve`, :func:`torch.eig`, :func:`torch.eig`,
+:func:`torch.symeig`. If any of the other ops would help your use case, please search if an issue has
+already been filed (https://github.com/pytorch/pytorch/issues?q=is%3Aopen+is%3Aissue+label%3A%22module%3A+named+tensor%22)
```
I think this is the wrong link no?

anjali411 replied:
yeah 😅 updated!

albanD commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,188 @@
+
+PyTorch supports Autograd for Complex Tensors.
+
+1. :func:`torch.functional.backward` can be used for holomorphic :math:`C -> C` functions.
```
You mean `torch.autograd.backward` right?
`torch.autograd.grad` can be used the same way (and the functional API as well) so maybe you want to be broader here and say something like "The autograd APIs"?

anjali411 replied:
yeah good idea! updated

albanD commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,188 @@
+   For non-holomorphic functions, the gradient is evaluated as if it were holomorphic.
+2. :func:`torch.functional.backward` can be used to optimize :math:`C -> R` functions, like
+   real-values loss functions of complex parameters :math:`x` by taking steps in the direction
+   of conjugate of :math:`x.grad`.
```
We might want to double check the conjugate here!
Also if this is true, we might want to update our optimizers to do that when the parameters are complex

anjali411 replied:
yeah that's correct! the gradient definition that we are using does specify that right?

albanD replied:
If it's already in the gradient definition, no need to do it again here.

albanD commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,188 @@
+
+Multiprocessing
+
+If any of these would help your use case, please search if an issue has already been filed (https://github.com/pytorch/pytorch/issues?q=is%3Aopen+is%3Aissue+label%3A%22module%3A+named+tensor%22)
```
Wrong link as well I think

anjali411 replied:
updated

ezyang replied:
Still looks wrong...

anjali411 replied:
yeah haven't pushed the local changes yet sorry 😅

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,188 @@
+
+Complex Numbers are numbers that can be expressed in the form :math:`a + bi`, where a and b are real numbers,
+and *i* is a solution of the equation :math:`x^2 = −1`. Complex numbers frequently occur in mathematics and
+engineering, especially in signal processing. The aim of introducing tensors of complex dtypes is to provide
```
Note that this documentation is going to live beyond its initial introduction, so you should try to talk about the feature as a timeless thing, and not like a release note.

anjali411 replied:
that's a very good point! Thanks Ed. I updated the doc but let me know if there are any more elements about it that are problematic

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,188 @@
+the real and imaginary values.
+
+We support many functions for Complex Tensors eg. :func:`torch.svd`, :func:`torch.qr` etc. Operations on complex tensors are likely to be
+faster than operations on float tensors mimicking them. Operations involving complex numbers in PyTorch are
```
faster and more memory efficient?

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,188 @@
+Angle and abs
+-------------
+
+The angle and absolute values of a complex tensor can be accesses using :func:`torch.angle` and
```
accessed

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,188 @@
+Linear Algebra
+--------------
+
+Currently, there is very minimal linear algebraic operation support for complex tensors.
```
"linear algebraic" isn't very idiomatic here. I'd probably personally write "there is very minimal support for linear algebra operations"

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,188 @@
+
+We do not support the following subsystems:
+
+Quantization
```
bullet list this?

ezyang commented:
what's the binary pt file?

mruberry commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,188 @@
+and *i* is a solution of the equation :math:`x^2 = −1`. Complex numbers frequently occur in mathematics and
+engineering, especially in signal processing. The aim of introducing tensors of complex dtypes is to provide
+a more natural user experience for users and libraries (eg. TorchAudio) that currently work around the
+lack of complex tensors by using Float Tensors with shape :math:`(..., 2)` where the last dimension contains
```
This seems like it's underselling the feature. The goal is to not just replace existing workarounds, right?

mruberry replied:
I would consider introducing these existing workarounds in its own section later. Maybe a section like: "adapting existing code"?

mruberry commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,188 @@
+          [-0.3773,  1.3487],
+          [-0.0861, -0.7981]])
+
+Accessing real and imag
```
This is really about acquiring just the real or imaginary part of the tensor, but the title seems like it's about the attributes themselves

anjali411 replied:
sorry I am not sure what you mean. can you provide an alternate suggestion?

anjali411 committed:
Update on "[WIP] Doc note for complex"




[ghstack-poisoned]

anjali411 commented:
> what's the binary pt file?

sorry that got added by mistake while I was just locally testing save and load. removed it!

anjali411 committed:
Update on "[WIP] Doc note for complex"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Doc note for complex"




[ghstack-poisoned]

anjali411 commented:
>> I'm not really sure why you want to compare users writing their own functions -- why would they do that?

@gchanan yeah good question. that's what a lot of libraries for complex have been doing right now. For eg. matrix multiplication of complex tensors (real tensors of shape `(...,2)`) are defined as functions by a some users and libraries to get the result expected if they were multiplying the equivalent complex tensors.

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,174 @@
+their own functions.
+
+.. note::
+     Spectral Ops currently don't use complex tensors but the API would be soon updated to use complex tensors.
```
nit: I think there's a tense disagreement; if you say "currently don't" then you should say "API will soon be updated"

ezyang replied:
nit: Spectral operations (don't abbreviate, don't capitalize)

ezyang replied:
nit: good to mention an e.g., list of what spectral operations are

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,174 @@
+-----------------------
+
+The real and imaginary values of a complex tensor can be accessed using the :attr:`real` and
+:attr:`imag` views.
```
nit: I think we talked about this in other parts of the documentation, but it is relying very hard on the reader's comprehension skills to only say "view" here, and expect them to unfold this into "accessing these attributes doesn't allocate any memory, and you can do in-place updates on them which will update the original complex tensor, and by the way, these tensors are not contiguous"

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,174 @@
+Accessing real and imag
+-----------------------
+
+The real and imaginary values of a complex tensor can be accessed using the :attr:`real` and
```
nit: It's probably more precise to say "Tensors containing only the real and imaginary values of a complex tensor"

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,174 @@
+Angle and abs
+-------------
+
+The angle and absolute values of a complex tensor can be accessed using :func:`torch.angle` and
```
nit: computed, not accessed

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,174 @@
+Serialization
+-------------
+
+Complex Tensors can be serialized, allowing data to be saved as complex values.
```
nit: tensors (no caps)

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,174 @@
+Autograd
+--------
+
+PyTorch supports Autograd for Complex Tensors. The autograd APIs can be
```
nit: autograd (no caps)

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,174 @@
+     >>> x.grad, y.grad
+     tensor([-0.6815+0.5931j,  0.5333-1.0872j]) tensor([-0.4869+0.9011j,  0.3673+0.2007j])
+     >>> x_.grad, y_.grad
+     tensortensor([-0.6815+0.5931j,  0.5333-1.0872j]) tensor([-0.4869+0.9011j,  0.3673+0.2007j])
```
looks like a copy paste bug tensortensor

anjali411 replied:
fixed! (also updated the example to make it more readable)

ezyang reviewed (approved):
okey

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,174 @@
+Complex numbers are numbers that can be expressed in the form :math:`a + bj`, where a and b are real numbers,
+and *j* is a solution of the equation :math:`x^2 = −1`. Complex numbers frequently occur in mathematics and
+engineering, especially in signal processing. Tensors of complex dtypes provide a more natural user experience
+for users and libraries (eg. TorchAudio) that previously worked around the lack of complex tensors by using
```
nit: the proper apa style for eg is "e.g.,"

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,174 @@
+and *j* is a solution of the equation :math:`x^2 = −1`. Complex numbers frequently occur in mathematics and
+engineering, especially in signal processing. Tensors of complex dtypes provide a more natural user experience
+for users and libraries (eg. TorchAudio) that previously worked around the lack of complex tensors by using
+float tensors with shape :math:`(..., 2)` where the last dimension contained the real and imaginary values.
```
nit: there still feels like there is some tense confusion here. Part of it is that there are plenty of libraries, TODAY, that represent complex as float tensors. So it should be present tense. Consider something like, "Tensors of complex dtypes provide a more natural user experience for working with complex numbers, compared with maintaining float tensors with shape :math:`(..., 2)` where the last dimension contains the real and imaginary values."

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,174 @@
+
+Operations on complex tensors (eg :func:`torch.mv`, :func:`torch.matmul`) are likely to be faster and more
+memory efficient than operations on float tensors mimicking them. Operations involving complex numbers in
+PyTorch are optimized to use vectorized assembly instructions and specialized kernels (e.g. LAPACK, CuBlas).
```
nit: official capitialization of cublas is cuBLAS

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,174 @@
+memory efficient than operations on float tensors mimicking them. Operations involving complex numbers in
+PyTorch are optimized to use vectorized assembly instructions and specialized kernels (e.g. LAPACK, CuBlas).
+Thus using functions for complex tensors will provide performance benefits as opposed to users defining
+their own functions.
```
nit: You probably don't actually need this sentence

anjali411 replied:
removed!

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,174 @@
+.. note::
+
+     The default dtype for complex tensors is determined by the default floating point dtype.
+     If the default floating point dtype is torch.float64 then complex numbers are inferred to
```
nit: backtick the code fragments

ezyang commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,174 @@
+
+PyTorch supports Autograd for Complex Tensors. The autograd APIs can be
+used for both holomorphic and non-holomorphic functions. For non-holomorphic
+functions, the gradient is evaluated as if it were holomorphic. For more details,
```
not sure if it's meaningful to say "evaluated as if it were holomorphic"; what does that even mean?

anjali411 replied:
ha I meant to say the gradient evaluation for non-holomorphic functions are done assuming the input function is holomorphic. I'll re-phrase to make it clear

anjali411 committed:
Update on "[WIP] Doc note for complex"




[ghstack-poisoned]

anjali411 committed:
Update on "[WIP] Doc note for complex"




[ghstack-poisoned]

gchanan commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,183 @@
+All factory functions apart from :func:`torch.linspace`, :func:`torch.logspace`, and :func:`torch.arange` are
+supported for complex tensors.
+
+Transition from the old representation
```
Can you consolidate all the historical information in this section?

anjali411 replied:
Yeah however I think some historical context in the introduction sets a better stage to truly justify `"more natural user experience"` and `"faster and more memory efficient than operations on float tensors mimicking them"`.

gchanan replied:
has anyone made the argument that mimicking complex numbers is a better user experience?  faster?  I haven't personally come across those arguments, my impression is that everyone recognizes the previous behavior as a hack.

gchanan commented on docs/source/complex_numbers.rst:
```diff
@@ -0,0 +1,183 @@
+::
+     >>> # computes the complex dot product for complex vectors
+     >>> # represented as float vectors
+     >>> # math: for complex numbers a and b vdot(a, b) = a.conj() * b
```
torch.sum(a.conj() * b)) ?

anjali411 replied:
synced offline: remove the code block from autograd section until we have a clear definition of functions that can be used with `view_as_real` and `view_as_complex` and for which we can ensure correct gradient computation.

anjali411 committed:
Update on "[WIP] Doc note for complex"




[ghstack-poisoned]

anjali411 committed:
Update on "Doc note for complex"


Differential Revision: [D22553266](https://our.internmc.facebook.com/intern/diff/D22553266)

[ghstack-poisoned]
